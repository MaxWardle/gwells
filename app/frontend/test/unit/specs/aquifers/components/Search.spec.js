/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import { shallowMount, createLocalVue } from '@vue/test-utils'
import SearchComponent from '@/aquifers/components/Search.vue'
import Vuex from 'vuex'
import axios from 'axios'
import VueRouter from 'vue-router'
import auth from '@/common/store/auth.js'

jest.mock('axios')

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

const aquiferFixture = {
  aquifer_id: '4',
  aquifer_name: 'Aggasiz',
  location_description: 'Aggasiz/Sea Bird Island',
  material_description: 'Sand and Gravel',
  litho_stratographic_unit: 'Fraser River Sediments',
  subtype_description: 'Unconfined sand and gravel - large river system',
  vulnerability_description: 'High',
  area: '63.3',
  productivity_description: 'High',
  demand: 'Low',
  mapping_year: '1993',
  resources: []
}

describe('Search Component', () => {
  const component = options =>
    shallowMount(SearchComponent, {
      localVue,
      stubs: ['router-link', 'router-view'],
      router: new VueRouter(),
      store: new Vuex.Store({
        modules: { auth }
      }),
      methods: {
        scrollToTableTop() {},
        fetchResourceSections() {
          this.aquifer_resource_sections = [
            { text: 'aquifer section', value: 'a' }
          ]
        }
      },
      ...options
    })

  it('has child checkboxes', () => {
    axios.get.mockResolvedValue({})

    expect(component({}).findAll('[type="checkbox"]').length).toEqual(1)
  })

  it('Displays a message if no aquifers could be found', () => {
    axios.get.mockResolvedValue({})
    const wrapper = component({
      computed: {
        aquiferList() {
          return []
        },
        emptyResults() {
          return true
        }
      },
      methods: {
        fetchResults() {}
      }
    })

    expect(wrapper.text()).toContain('No aquifers could be found')
  })

  it('Matches the snapshot', () => {
    const wrapper = component({
      computed: {
        aquiferList() {
          return [aquiferFixture]
        },
        emptyResults() {
          return false
        },
        displayPageLength() {
          return 0
        }
      },
      methods: {
        fetchResults() {}
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('search updates route query params', () => {
    axios.get.mockResolvedValue({})

    const wrapper = component()

    wrapper.find('#aquifers-search-field').setValue('asdf')

    wrapper.find('input[type="checkbox"]').setChecked()
    wrapper.find('form').trigger('submit')

    expect(axios.get).toHaveBeenCalledWith('aquifers', {
      params: { resources__section__code: 'a', search: 'asdf' }
    })
  })

  it('form reset resets response and query', () => {
    const wrapper = component()
    axios.get.mockResolvedValue({})

    wrapper.find('#aquifers-search-field').setValue('asdf')
    wrapper.find('form').trigger('submit')

    expect(wrapper.vm.query.search).toEqual('asdf')

    wrapper.find('form').trigger('reset')

    expect(wrapper.vm.query.search).toEqual(undefined)
  })
})
