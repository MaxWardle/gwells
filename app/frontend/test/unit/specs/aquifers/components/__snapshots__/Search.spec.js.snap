// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Matches the snapshot 1`] = `
<div
  class="container container-wide p-1"
>
   
  <div
    class="card p-3 mb-4"
  >
    <!---->
    <!---->
    <h1>
      Aquifer Search
    </h1>
     
    <div
      class="pb-2"
    >
      <!---->
    </div>
     
    <!---->
     
    <form
      class=""
    >
      <div
        class="form-row"
      >
        <div
          class="col-md-4 col-12"
        >
          <fieldset
            class="form-group"
          >
            <legend
              class="col-form-label pt-0"
              tabindex="-1"
            >
              Aquifer number
            </legend>
            <div
              role="group"
              tabindex="-1"
            >
              <input
                class="form-control"
                id="aquifers-number"
                type="text"
              />
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
           
          <fieldset
            class="form-group"
          >
            <legend
              class="col-form-label pt-0"
              tabindex="-1"
            >
              Aquifer name
            </legend>
            <div
              role="group"
              tabindex="-1"
            >
              <input
                class="form-control"
                id="aquifers-name"
                type="text"
              />
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
        </div>
      </div>
       
      <div
        class="form-row"
      >
        <div
          class="col"
        >
          <fieldset
            class="form-group"
          >
            <!---->
            <div
              role="group"
              tabindex="-1"
            >
              <button
                class="btn btn-primary"
                id="aquifers-search"
                type="submit"
              >
                Search
              </button>
               
              <button
                class="btn btn-default"
                type="reset"
              >
                Reset
              </button>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
        </div>
      </div>
    </form>
     
    <table
      aria-busy="false"
      aria-colcount="11"
      class="table b-table mt-3 table-striped"
      id="aquifers-results"
    >
      <!---->
      <!---->
      <thead
        class=""
      >
        <!---->
        <tr>
          <th
            aria-colindex="1"
            aria-label="Click to sort Descending"
            aria-sort="ascending"
            class=""
            scope="col"
            tabindex="0"
          >
            Aquifer number
          </th>
          <th
            aria-colindex="2"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Aquifer name
          </th>
          <th
            aria-colindex="3"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Descriptive location
          </th>
          <th
            aria-colindex="4"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Material
          </th>
          <th
            aria-colindex="5"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Litho stratigraphic unit
          </th>
          <th
            aria-colindex="6"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Subtype
          </th>
          <th
            aria-colindex="7"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Vulnerability
          </th>
          <th
            aria-colindex="8"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Size-kmÂ²
          </th>
          <th
            aria-colindex="9"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Productivity
          </th>
          <th
            aria-colindex="10"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Demand
          </th>
          <th
            aria-colindex="11"
            aria-label="Click to sort Descending"
            aria-sort="none"
            class=""
            scope="col"
            tabindex="0"
          >
            Year of mapping
          </th>
        </tr>
      </thead>
      <!---->
      <tbody
        class=""
      >
        <!---->
        <tr
          class=""
        >
          <td
            aria-colindex="1"
            class=""
          >
            <router-link-stub
              to="[object Object]"
            />
          </td>
          <td
            aria-colindex="2"
            class=""
          >
            Aggasiz
          </td>
          <td
            aria-colindex="3"
            class=""
          >
            Aggasiz/Sea Bird Island
          </td>
          <td
            aria-colindex="4"
            class=""
          >
            
        Sand and Gravel
      
          </td>
          <td
            aria-colindex="5"
            class=""
          >
            Fraser River Sediments
          </td>
          <td
            aria-colindex="6"
            class=""
          >
            
        Unconfined sand and gravel - large river system
      
          </td>
          <td
            aria-colindex="7"
            class=""
          >
            
        High
      
          </td>
          <td
            aria-colindex="8"
            class=""
          >
            63.3
          </td>
          <td
            aria-colindex="9"
            class=""
          >
            
        High
      
          </td>
          <td
            aria-colindex="10"
            class=""
          >
            
        
      
          </td>
          <td
            aria-colindex="11"
            class=""
          >
            1993
          </td>
        </tr>
        <!---->
        <!---->
      </tbody>
    </table>
     
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          
          Showing 1 to 0 of 
        
        </div>
         
        <!---->
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
